src/workerd/api/web-socket.h

```diff
@@ -180,6 +244,102 @@ class WebSocket: public EventTarget {
   // object to be kept live while waiting for the promise returned by couple() to complete. Instead,
   // the promise takes direct ownership of the underlying KJ-native WebSocket (as well as `other`).
 
+  kj::Own<kj::WebSocket> acceptAsHibernatable() {
+    // Extract the kj::WebSocket from this api::WebSocket (if applicable). The kj::WebSocket will be
+    // owned elsewhere, but the api::WebSocket will retain a reference.
+    KJ_IF_MAYBE(hibernatable, farNative->state.tryGet<AwaitingAcceptanceOrCoupling>()) {
+      // We can only request hibernation if we have not called accept.
+      auto ws = kj::mv(hibernatable->ws);
+      // We pass a reference to the kj::WebSocket for the api::WebSocket to refer to when calling
+      // `send()` or `close()`.
+      farNative->state.init<Accepted>(
+          Accepted::Hibernatable{ .ws = *ws }, *farNative, IoContext::current());
+      return kj::mv(ws);
+    }
+    JSG_FAIL_REQUIRE(TypeError,
+        "Tried to make an api::WebSocket hibernatable when it was in an incompatible state.");
+  }
+
+  inline void tryReleaseNative(jsg::Lock& js) {
+    // If the native WebSocket is no longer needed (the connection closed) and there are no more
+    // messages to send, we can discard the underlying connection.
+    auto& native = *farNative;
+    if ((native.closedOutgoing || native.outgoingAborted) && !native.isPumping) {
+      // Native WebSocket no longer needed; release.
+      KJ_ASSERT(native.state.is<Accepted>());
+      native.state.init<Released>();
+    }
+  }
+
+  enum HibernatableReleaseState {
```

Comment:
Nit: 
```suggestion
  enum class HibernatableReleaseState {
```

Reply:
Didn't know there was a difference, thanks for mentioning!